File: Programmers/Algorithm_Kit/07_dynamic_programming/002.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm> //max_element
using namespace std;

int dp[501][501];

int solution(vector<vector<int>> triangle) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    
    int answer = 0;

    dp[1][1] = triangle[0][0];

    int n = triangle.size();

    for (int i = 2; i <= n; i++) {
        for (int j = 1; j <= i; j++) {
            dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - 1]) + triangle[i - 1][j - 1];
        }
    }

    answer = *(max_element(dp[n] + 1, dp[n] + n + 1));

    return answer;
}

File: programmers/algorithmkit/07_dynamic_programming/002.java
import java.util.*;

class Solution {
    int[][] dp = new int[501][501];

    public int solution(int[][] triangle) {
        int answer = 0;
        
        int n = triangle.length;
        
        dp[1][1] = triangle[0][0];

        for (int i = 2; i <= n; i++) {
            for (int j = 1; j <= i; j++) {
                dp[i][j] = Math.max(dp[i - 1][j - 1], dp[i - 1][j]) + triangle[i - 1][j - 1];
            }
        }

        for (int i = 1; i <= n; i++) {
            answer = Math.max(dp[n][i], answer);
        }

        return answer;
    }
}
